jobs/kpi-forecasting/kpi_forecasting/models/funnel_forecast.py

```diff
@@ -17,13 +32,590 @@ class SegmentModelHolder:
+                raise ValueError(
+                    "All forecast_df columns except 'submission_date' and segment dims must be numeric,"
+                    f" but column {i} has type {df[i].dtypes}."
+                )
+
+    def _percentile_name_map(self, percentiles: List[int]) -> Dict[str, str]:
+        percentiles.sort()
+        return {
+            f"p{percentiles[0]}": "value_low",
+            f"p{percentiles[1]}": "value_mid",
+            f"p{percentiles[2]}": "value_high",
+            "mean": "value",
+        }
+
+    def _summarize(
+        self,
+        segment_settings: SegmentModelSettings,
+        period: str,
+        numpy_aggregations: List[str],
+        percentiles: List[int],
+    ) -> pd.DataFrame:
+        """
+        Calculate summary metrics for `forecast_df` over a given period, and
+        add metadata.
+        """
+        if len(percentiles) != 3:
+            print(
+                """
+                  Can only pass a list of length 3 as percentiles, for lower, mid, and upper values.
+                  Will default to using [10, 50, 90].
+                  """,
+                flush=True,
+            )
+            percentiles = [10, 50, 90]
+        # build a list of all functions that we'll summarize the data by
+        aggregations = [getattr(np, i) for i in numpy_aggregations]
+        aggregations.extend([pdx.percentile(i) for i in percentiles])
+
+        # aggregate metric to the correct date period (day, month, year)
+        observed_summarized = pdx.aggregate_to_period(
+            (
+                self.observed_df.loc[
+                    (
+                        (
+                            self.observed_df[list(segment_settings.segment)]
+                            == pd.Series(segment_settings.segment)
+                        ).all(axis=1)
+                    )
+                    & (
+                        self.observed_df["submission_date"]
+                        >= datetime.strptime(
+                            segment_settings.start_date, "%Y-%m-%d"
+                        ).date()
+                    )
+                ].copy()
+            ),
+            period,
+        )
+        forecast_agg = pdx.aggregate_to_period(segment_settings.forecast_df, period)
+
+        # find periods of overlap between observed and forecasted data
+        overlap = forecast_agg.merge(
+            observed_summarized,
+            on="submission_date",
+            how="left",
+        ).fillna(0)
+
+        forecast_summarized = (
+            forecast_agg.set_index("submission_date")
+            # Add observed data samples to any overlapping forecasted period. This
+            # ensures that any forecast made partway through a period accounts for
+            # previously observed data within the period. For example, when a monthly
+            # forecast is generated in the middle of the month.
+            .add(overlap[["value"]].values)
```

Comment:
Why is it being added here?  I don't follow the comment

Reply:
It's possible for the forecast and observed data to overlap when you aggregate, e.g. if you're forecasting for June and starting with today's date, your monthly aggregation would require those first few days of June from the historical data to be correct. This will take those days from the historical data and add them to the June forecast

Reply:
Okay I think I get it.  Will the overlap be at most one period, or could you have a larger overlap and if so why does this still work?

Reply:
And why do you need to additional aggregation step after this happens?

Reply:
I think the overlap could only be one period, yeah. I can't think of an instance where it wouldn't be. The aggregation after this will take those days that overlap include them when aggregating to the month.

Reply:
Is the second agg redundant?  Both `forecast_agg` and `observed_summarized` are aggregated to `submission_date`, right?

Reply:
They are, yeah, but this will re-perform the monthly aggregations. So when it adds in the overlap rows to the forecast set, from the historical set, that overlapping month will have 2 rows in the DataFrame. The second agg call aggregates that overlap.

Reply:
You might be able to simplify it by concat-ing both unaggregated dataframes and then doing the agg?  But maybe that is harder to read?  idk if it's worth messing around with but this is one of the knottier bits of code in this module (which is indicative of how clearly the written the rest of it is!)

Reply:
Sounds good to me, resolving